Add tune resonances plot

oscarxblanco opened the pull request:
Dear all,
this PR is a work in progress.
It implements a tune resonance plot mentioned in the discussion #808 .
Here is an example of the default plot.
```python
import at, at.plot
at.plot.standalone.plot_tune2D_resonances()
```
![image](https://github.com/user-attachments/assets/0827c26b-3fc0-4ec2-bc16-4f8b59e25d2e)
Here is an example with fifth orders resonances in symmetry 7.
```python
at.plot.standalone.plot_tune2D_resonances(orders=[5],period=7,window=numpy.array([71 ,73, 45, 48]))
```
![image](https://github.com/user-attachments/assets/1f48b11a-888d-4ddd-aea3-07863aa8c6c6)
This is similar  to the result of `reson` from MML.
```matlab
reson(5,7,[71 73 45 48])
```
![image](https://github.com/user-attachments/assets/1667c697-e75b-4a31-8186-7b399fe4d29a)



I think this could be useful, would you consider to add it @swhite2401 ?
I have put it for the moment in at.plot.standalone. Is this the right place @swhite2401 ?

lfarv commented:
Hi @oscarxblanco,
Nice addition ! I think that `at.plot.standalone` becomes larger and larger, with functions unrelated to each other. Why don't you put your function in its own module?

swhite2401 commented:
@lfarv, sorry to go a bit off topic but it has always been assumed that importing matplolib introduces significant performance degradation.
This in return introduces counter intuitive import a plot libraries, that have already been raised by @lnadolski (can't find the discussion back...).
The question is: has this been evaluated? All the other code I can think off import matplotlib at initialization so I am wondering if we are not adding complexity for marginal gain...

lfarv commented:
@swhite2401:  There are in fact 2 questions raised by your comment:

1. `matplotlib` is an **optional** dependency of PyAT. You don't get it installed with:
   ```python
   pip install accelerator-toolbox
   ```
   But you get it installed with:
   ```python
   pip install "accelerator-toolbox[plot]"
   ```
   This is intended to minimise the dependencies when PyAT is installed in a virtual environment to power a virtual accelerator, for instance. Here is a comparison of the installation of PyAT in a fresh virtual environment, with and without the `[plot]` modifier:
   - without plot
     ```
     (noplot)  laurent@garou ~ % pip install accelerator-toolbox
     ...
     (noplot)  laurent@garou ~ % pip list
     Package             Version
     ------------------- -------
     accelerator-toolbox 0.6.1
     numpy               2.0.1
     pip                 24.1.2
     scipy               1.14.0
     setuptools          70.3.0
     (noplot)  laurent@garou ~ % 
     ```
   - with plot
     ```
     (plot)  laurent@garou ~ % pip install "accelerator-toolbox[plot]"
     ...
     (plot)  laurent@garou ~ % pip list
     Package             Version
     ------------------- -------
     accelerator-toolbox 0.6.1
     contourpy           1.2.1
     cycler              0.12.1
     fonttools           4.53.1
     kiwisolver          1.4.5
     matplotlib          3.9.0
     numpy               2.0.1
     packaging           24.1
     pillow              10.4.0
     pip                 24.1.2
     pyparsing           3.1.2
     python-dateutil     2.9.0.post0
     scipy               1.14.0
     setuptools          70.3.0
     six                 1.16.0
     ```
   This makes sense and can be kept, but the installation instructions should be improved to explain the use of the `[plot]` modifier.


2. `at.plot` is **not automatically imported** when importing `at`. Here the only question is the import time. I agree that this should not be a problem, and that we could import `at.plot` by default.

So my proposal is:
- keep the installation process as it is, with improved documentation for the `[plot]`, `[dev]`, `[mpi]` and `[doc]` modifiers,
- import `at.plot` by default. If `matplotlib` has not been initially installed, this just prints a message saying: _'Plotting functions require matplotlib, to use them run "pip install matplotlib"'_. This message is harmless for a non-graphics installation, and has no time penalty.

swhite2401 commented:
@lfarv, if we allow importing matplotlib, each module could now host its own plotting function, right? Then we don't have to pile-up things in `at.plot.standalone`.
Maybe this PR could be used to start this new organization?

oscarxblanco commented:
Dear @lfarv and @swhite2401 ,
this PR is ready for review, who should do it?
I have also included the change you mentioned to import at.plot by default.

lfarv commented:
@oscarxblanco 

>  if we allow importing matplotlib, each module could now host its own plotting function, right? Then we don't have to pile-up things in at.plot.standalone.
> Maybe this PR could be used to start this new organization?

You are right, there is no need to pile up things in `standalone.py`. The point is to keep all graphics functions in a single package `at.plot`, so that a single check for the presence of matplotlib in `at.plot.__init__.py` is enough to skip the whole package. No need to check in each module.

Concerning this PR, it's better keeping it centered on the resonance plot, as it is, avoiding mixing other topics. I would have put the modification of `at.__init__.py` in another PR, as discussed above. But since it's done, keep it, don't worry !

lfarv commented on pyat/at/plot/resonances.py:
```diff
@@ -0,0 +1,505 @@
+
+
+def plot_tune2d_resonances(
+    **kwargs: dict[str, any],
```
Most of the keyword arguments have an immutable default value (period, verbose, includelegend, block,…). In such a case, it's better to declare them in the function declaration, if possible with a type hint. Then they appear with their type and default value in the documentation, in the signature of the function.

oscarxblanco replied:
Done.

lfarv commented on pyat/at/plot/resonances.py:
```diff
@@ -0,0 +1,505 @@
+    9th: 'lightgreen', width 1
+    10th to 15th: RGB increased in steps of 0.1, width 1
+
+    Custom Style:
+    You could pass a custom line style in a dictionary as
+    custom_linesty = mydictionary,
+    where mydictionary should contain two entries
+    dict(0: prop_normal, 1: prop_skew).
+    prop_normal and prop_skew are also dictionaries starting at zero,
+    i.e. the_nth_order-1. Each entry contains the line properties of the nth-1
+    resonance to plot.
```
Why not simply `linestyles` for the keyword? (by the way, in the code, you use `customlinesty` without underscore)

Then you could use "normal" and "skew" for the keys of the top-level directory, and "order" instead of "order-1" for the keys of the subdirectories;

oscarxblanco replied:
Done.

lfarv commented on pyat/at/plot/resonances.py:
```diff
@@ -0,0 +1,505 @@
+        13: prop14s.copy(),
+        14: prop15s.copy(),
+    }
+    # assemble final dict
+    defaultlprop = {0: propn.copy(), 1: props.copy()}
```
The default lifestyle directory could be initialized at the module level, so that it's done once for ever at import time. And are all the `copy()` necessary ?
However, it's a detail, I don't care too much…

oscarxblanco replied:
All call to copies where removed. They were not really needed.
Also, the dictionary was moved out of the function.

lfarv commented:
I'd suggest to simplify the keywords:
- "legend" instead of "includelegend"
- "linestyles" instead of "custom_linesty"
- "debug" instead of "debugverbose"

Any advice, @swhite2401 ?

swhite2401 commented:
> I'd suggest to simplify the keywords:
> 
>     * "legend" instead of "includelegend"
> 
>     * "linestyles" instead of "custom_linesty"
> 
>     * "debug" instead of "debugverbose"
> 
> 
> Any advice, @swhite2401 ?

I agree on the simpler / shorter keywords. What @lfarv proposes looks fine to me

lfarv commented:
And what about `plot_tune_diagram` ?

I have another question about systematic and non-systematic resonances. As the function is now, you have to alternate 2 plots:
- if you set period = _machine_perodicity_,  you get only the systematic resonances,
- if you set period = 1, you get both systematic and non-systematic ones, but you cannot distinguish systematic ones.
Could there be a line property allowing to enhance systematic resonances? The line width is already used for the order…

Anyway, if there is no obvious solution, forget it, we take the function as it is, it can be improves later !

oscarxblanco commented:
> @oscarxblanco
> 
> > if we allow importing matplotlib, each module could now host its own plotting function, right? Then we don't have to pile-up things in at.plot.standalone.
> > Maybe this PR could be used to start this new organization?
> 
> You are right, there is no need to pile up things in `standalone.py`. The point is to keep all graphics functions in a single package `at.plot`, so that a single check for the presence of matplotlib in `at.plot.__init__.py` is enough to skip the whole package. No need to check in each module.
> 
> Concerning this PR, it's better keeping it centered on the resonance plot, as it is, avoiding mixing other topics. I would have put the modification of `at.__init__.py` in another PR, as discussed above. But since it's done, keep it, don't worry !

Dear @lfarv , I have removed the .plot import from __init__.py. I think you are right about having a separted PR because it will be clear when the change was introduced.

oscarxblanco commented:
> I'd suggest to simplify the keywords:
> 
>     * "legend" instead of "includelegend"
> 
>     * "linestyles" instead of "custom_linesty"
> 
>     * "debug" instead of "debugverbose"
> 
> 
> Any advice, @swhite2401 ?

Done.

oscarxblanco commented:
> And what about `plot_tune_diagram` ?
> 
> I have another question about systematic and non-systematic resonances. As the function is now, you have to alternate 2 plots:
> 
>     * if you set period = _machine_perodicity_,  you get only the systematic resonances,
> 
>     * if you set period = 1, you get both systematic and non-systematic ones, but you cannot distinguish the systematic ones.
>       Could there be a line property allowing to enhance systematic resonances? The line width is already used for the order…
> 
> 
> Anyway, if there is no obvious solution, forget it, we take the function as it is, it can be improves later !

Dear @lfarv ,
maybe something like this ?
systematic resonances in variable width, normal are continuous (-), skew are dashed (- -).
non-systematic resonances in width=1, normal are dashdot (-.), skew are dotted (. . . ).
![image](https://github.com/user-attachments/assets/96e6a339-5fb3-48fd-9b63-c2657d728a91)

oscarxblanco commented:
Another example,
![image](https://github.com/user-attachments/assets/4aa76b4c-2376-4b7e-a844-644be6ccde3b)

lfarv commented on pyat/at/plot/resonances.py:
```diff
@@ -0,0 +1,505 @@
+    15: prop15s,
+}
+# assemble final dict
+defaultlprop = {0: propn, 1: props}
```
Still: I think it would be more clear to have
```python
defaultlprop = {"normal": propn, "skew": props}
```

oscarxblanco replied:
Done.

lfarv commented:
Your last plots look nice: that would be perfect.

A last detail: your 'period' argument corresponds to the `periodicity` attribute of the lattice object. Could you switch to the same name? Sorry to intervene on such details, but I'm trying to get uniform naming across all modules…

I tried your function in a Jupyter notebook, and I get a strange thing: when validating the cell, I get two identical figures instead of one. I've never seen this. Did you try that? Could it be a problem with my setup?

oscarxblanco commented:
Dear @lfarv ,
it sounds like something related to this thread
https://stackoverflow.com/questions/62322083/why-on-calling-plt-figure-in-jupyter-notebook-two-figure-appear

Wrt the names, there is no problem, I will do the changes.

I think that it might be better to have the option to use axes created by the user, similar to synoptics. It will allow to plot and and the lines.

oscarxblanco commented:
> Your last plots look nice: that would be perfect.
> 
> A last detail: your 'period' argument corresponds to the `periodicity` attribute of the lattice object. Could you switch to the same name? Sorry to intervene on such details, but I'm trying to get uniform naming across all modules…
> 
> I tried your function in a Jupyter notebook, and I get a strange thing: when validating the cell, I get two identical figures instead of one. I've never seen this. Did you try that? Could it be a problem with my setup?

I did the change about periodicity.
Wrt to jupyter I will have to install it and test.

oscarxblanco commented:
> Your last plots look nice: that would be perfect.

Dear @lfarv , I have modified the function to overlay plots using the same axes, and I have added options to slightly modify the default line properties to avoid the need for the user to create another complete custom dictionary.

With those modifications one could overlay the plot in symmetry $n$ to a plot in symmetry $1$.
Here is an example using dots, width of 1, and black color.
```python
>>> import at, at.plot
>>> ax = at.plot.resonances.plot_tune_diagram(periodicity=3,window=[0,0.45,0,0.48])
>>> ax2 = at.plot.resonances.plot_tune_diagram(axes=ax,linewidth=1,linecolor='k',linestyle='dots')
```
![image](https://github.com/user-attachments/assets/d152d2b3-705f-416c-bb82-2a58f1ad1186)
And here another example, the one I showed yesterday with only third order resonances.
```python
ax = at.plot.resonances.plot_tune_diagram(orders=3,periodicity=3,window=[23,23.5,40,40.5])
vv = at.plot.resonances.plot_tune_diagram(orders=3,axes=ax,linewidth=1,linestyle='dots')
```
![image](https://github.com/user-attachments/assets/32c19868-570c-4008-80bd-94805843de5f)

oscarxblanco commented:
Dear @lfarv ,
I have tried to reproduce the behavior of the double plot you see using jupiterlab. In fact, I don't get a double plot but I have the impression that jupyterlab creates its own figure and forces the plots to be visible, ignoring what is in the function.

I manage workaround it by adding another flag called `show` set to `True` by default in order to get the plot in any python3 session, and should be set to `False` when used in jupyterlab. When working in jupyterlab one should create the figure, set the axes limits and then use those axes for the plot.

Here is a screenshot and the code:
```python
fig = plt.figure();
ax = fig.add_subplot(111);
ax.set_xlim([0,0.45]);
ax.set_ylim([0,0.45]);
ax, hs, ls = at.plot.resonances.plot_tune_diagram(orders=3,axes=ax,periodicity=3,window=[23,23.5,40,40.5],legend=True,addtolabel=" syst.",show=False);
vx, h, l = at.plot.resonances.plot_tune_diagram(orders=3,axes=ax,linewidth=1,linestyle='dots',legend=True, addtolabel=" non syst.",show=False);
```
![image](https://github.com/user-attachments/assets/b684da38-26df-4b75-a777-975b8b4e8ed4)

lfarv commented on pyat/at/plot/resonances.py:
```diff
@@ -0,0 +1,420 @@
+    linecolor = kwargs.pop("linecolor", None)
+    linewidth = kwargs.pop("linewidth", None)
+    addtolabel = kwargs.pop("addtolabel", None)
+    defaultlprop = create_linepalette(
+        linestyle=linestyle,
+        linecolor=linecolor,
+        linewidth=linewidth,
+        addtolabel=addtolabel,
+    )
+    lprop = kwargs.pop("linedict", defaultlprop)
```
A suggestion (forget it if you find it irrelevant):
```python
    lprop = create_linepalette(
         linestyle=linestyle,
         linecolor=linecolor,
         linewidth=linewidth,
         addtolabel=addtolabel,
     )
    userprop = kwargs.pop("linedict", {})
    lprop["normal"].update(userprop.get("normal", {}))
    lprop["skew"].update(userprop.get("skew", {}))
```
The present behavior is kept, but it allows the user to overwrite only the lines he wants to change, others are unchanged.

oscarxblanco replied:
Done.

lfarv commented on pyat/at/plot/resonances.py:
```diff
@@ -0,0 +1,420 @@
+    maxy = maxy + periodicity - numpy.mod(maxy, periodicity)
+    minmaxydist = maxy - miny
+
+    linestyle = kwargs.pop("linestyle", None)
+    linecolor = kwargs.pop("linecolor", None)
+    linewidth = kwargs.pop("linewidth", None)
+    addtolabel = kwargs.pop("addtolabel", None)
```
These could also be moved to the function declaration (better documentation).

oscarxblanco replied:
Done.

lfarv commented:
> I have tried to reproduce the behavior of the double plot you see using jupiterlab. In fact, I don't get a double plot but I have the impression that jupyterlab creates its own figure and forces the plots to be visible, ignoring what is in the function.

Sorry the my late reply: what I get is described in the link you posted: if I don't store the function return value in a variable, Jupyter notebook displays the function result, which creates another figure. This is the normal notebook behavior. By the way, it disappeared with the new return values. So your "show" keyword is not necessary, though it can be kept, of course.

lfarv reviewed (approved):
OK for me, nice addition!

swhite2401 commented:
Hell @oscarxblanco ,
On my python3.8 environment at crashes with the following error:
![image](https://github.com/user-attachments/assets/7a4e7d14-f034-434d-aaee-c60bbb2b9bc2)

it seems ` from __future__ import annotations` is missing in `resonance.py`, at least it fixes the problem for me.

How could this pass the tests? This is a bit puzzling to me...

Should I revert and implement the fix?

oscarxblanco commented:
I'll fix it. I don't know how it passed that test. I work on python 3.11.2 and didn't notice it.
